WalletWasabi/WabiSabi/Backend/Rounds/Round.cs

```diff
@@ -120,5 +123,18 @@ public bool IsInputRegistrationEnded(uint maxInputCount, TimeSpan inputRegistrat
 
 			return false;
 		}
+
+		public bool TryAddMoreWeight(uint inputWeightSum)
+		{
+			var toSubstract = RegistrableWeightCredentials - inputWeightSum;
+
+			if (RemainingWeightUnits < toSubstract)
+			{
+				return false;
+			}
+
+			RemainingWeightUnits -= toSubstract;
```

Comment:
if `CredentialIssuer.HandleRequest` or `RemoveDuplicateAlices` throws an error after `TryAddMoreWeight` is called the subtracted amount here won't be added back (unauthorized clients could drain the remaining WU allocation var).

Since the arena has a global mutex throughout the action, I think the correct approach is to have the check happen before, but only subtract when `Alices.Add` happens at line 141.

The arena could just compare `RemainingWeightUnits` in its condition at line 130, and an unmodified `TryAddMoreWeight` would be called last thing before `Alices.Add`, its name already implies it mutates state.

Reply:
Yes, I think that's cleaner. `Round.cs` represents the backend's state for the round and since we now have two dependent properties we need to keep consistent, it makes sense for it to have an API that defines the operations that transition states on it, to do that that `Arena` and `InputRegistrationHandler` can just use them.

`Arena` should be responsible for the server state as a whole handling requests by updating the individual `Round`s while ensuring consistency between them (input can only go into one round), and enforcing banning.

The input registration has a mixed main role right now, the first part is is to resolve `OutPoint` to a `Coin` and implementing the banning checks before the `Arena` is locked for the actual input registration handling, and its second responsibility is to do all of the `Round` state management that `Arena` does for other handlers.

I've been working on refactoring some of `Round.cs` into a shared model object (which will be in a property of `Round.cs` as well as a future client counterpart of that). This object contains the blockchain level checks that `Arena` and `InputRegistrationHandler` do (e.g. script types, witness validation...), since they are needed on both the clients and the backend, so we can expect these to shrink even more (I haven't done a PR yet, main bottleneck has been learning a few various C# features and limitations, i've been trying different approaches...).

After that PR if the inner (locked) parts of InputRegistrationHandler bits have shrunk enough I think we can just move them and the banning logic to `Arena`. Then the remaining parts of `InputRegistrationHandler` are the preprocessing that allows `Arena` to always work with validated (confirmed, unspent) `Coin` objects, that can probably be renamed to something like `OutPointResolver` or something?